File: scripts/coverage-diff.mjs
import { readFileSync, existsSync } from "node:fs";
import { join } from "node:path";
import { fileURLToPath } from "node:url";

export function loadReport(dir, label) {
  const path = join(dir, `${label}-report.json`);
  return existsSync(path) ? JSON.parse(readFileSync(path, "utf-8")) : null;
}

export function pct(v) {
  return v != null ? `${v.toFixed(1)}%` : "—";
}

export function deltaIcon(d) {
  if (d == null) return "";
  if (d < -0.5) return " 🔻";
  if (d > 0.5) return " ✅";
  return "";
}

function main() {
  const [currentDir, baselineDir] = process.argv.slice(2);
  if (!currentDir || !baselineDir) {
    console.error("Usage: coverage-diff.mjs <current-dir> <baseline-dir>");
    process.exit(1);
  }

  const scopes = ["go", "node", "mobile"];
  const labels = { go: "Go API", node: "api-client", mobile: "Mobile" };
  const md = ["### Coverage Report", "", "| Scope | Baseline | Current | Change |", "|---|---|---|---|"];

  for (const scope of scopes) {
    const cur = loadReport(currentDir, scope);
    const base = loadReport(baselineDir, scope);
    if (!cur && !base) continue;

    const cv = cur?.total;
    const bv = base?.total;
    if (cv != null && bv != null) {
      const d = cv - bv;
      md.push(`| **${labels[scope]}** | ${pct(bv)} | ${pct(cv)} | ${d > 0 ? "+" : ""}${d.toFixed(1)}%${deltaIcon(d)} |`);
    } else {
      md.push(`| **${labels[scope]}** | ${pct(bv)} | ${pct(cv)} | — |`);
    }
  }

  md.push("", "> _Non-blocking coverage report._");
  process.stdout.write(md.join("\n") + "\n");
}

if (process.argv[1] === fileURLToPath(import.meta.url)) {
  main();
}

File: magefile/coverage_diff.go
package main

import (
	"fmt"
	"strings"
)

type CoverageReport struct {
	Total float64 `json:"total"`
}

func CoverageComment(current, baseline map[string]*CoverageReport) string {
	type scope struct {
		key   string
		label string
	}
	scopes := []scope{{key: "api", label: "Go API"}, {key: "bot", label: "Discord Bot"}}

	lines := []string{
		"### Coverage Report",
		"",
		"| Scope | Baseline | Current | Change |",
		"|---|---|---|---|",
	}

	for _, scope := range scopes {
		cur := current[scope.key]
		base := baseline[scope.key]
		if cur == nil && base == nil {
			continue
		}
		if cur != nil && base != nil {
			delta := cur.Total - base.Total
			lines = append(lines, fmt.Sprintf("| **%s** | %s | %s | %s%s |", scope.label, pct(base.Total), pct(cur.Total), signedPct(delta), deltaIcon(delta)))
			continue
		}
		lines = append(lines, fmt.Sprintf("| **%s** | %s | %s | — |", scope.label, nullablePct(base), nullablePct(cur)))
	}

	lines = append(lines, "", "> _Non-blocking coverage report._")
	return strings.Join(lines, "\n") + "\n"
}

func pct(v float64) string {
	return fmt.Sprintf("%.1f%%", v)
}

func nullablePct(report *CoverageReport) string {
	if report == nil {
		return "—"
	}
	return pct(report.Total)
}

func signedPct(delta float64) string {
	if delta > 0 {
		return fmt.Sprintf("+%.1f%%", delta)
	}
	return fmt.Sprintf("%.1f%%", delta)
}

func deltaIcon(delta float64) string {
	if delta < -0.5 {
		return " 🔻"
	}
	if delta > 0.5 {
		return " ✅"
	}
	return ""
}
